File: app/src/main/java/com/example/kmmkitchentmanagement/fragmenthome/frag_CongCu.java
package com.example.kmmkitchentmanagement.fragmenthome;

import android.content.Context;
import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.kmmkitchentmanagement.Model.LuanVan;
import com.example.kmmkitchentmanagement.R;
import com.example.kmmkitchentmanagement.fragmentSub.frag_appinforView;
import com.example.kmmkitchentmanagement.fragmentSub.frag_lichsu;
import com.example.kmmkitchentmanagement.interfaceFile.LogoutListener;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.AggregateQuerySnapshot;
import com.google.firebase.firestore.AggregateSource;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.QuerySnapshot;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;


public class frag_CongCu extends Fragment {
    FirebaseFirestore firestore;
    SwipeRefreshLayout refreshLayout;
    TextView all,recent7,recent30,onprogress,published,txtCCTongTG,txtCCTongCD;
    private Button mExit,historyBtn,ViewInforBtn,savetofileBtn;

    public frag_CongCu() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_frag__cong_cu, container, false);

        AddView(view);
        Exit(view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        FirebaseInit();
        DataHandler();
        EventHandler();
    }
    protected void FirebaseInit() {
        firestore=FirebaseFirestore.getInstance();
    }
    protected void DataHandler(){
        Date last7day=new Date();
        last7day.setDate(last7day.getDate()-7);
        Date last30day=new Date();
        last30day.setDate(last30day.getDate()-30);
        firestore.collection("/luanvan").count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        all.setText(String.valueOf(task.getResult().getCount()));
                    }
                });
        firestore.collection("/luanvan")
                .whereGreaterThan("addTime",last7day).count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        recent7.setText( String.valueOf(task.getResult().getCount()));
                    }
                });
        firestore.collection("/luanvan")
                .whereGreaterThan("addTime",last30day).count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        recent30.setText(String.valueOf(task.getResult().getCount()));
                    }
                });
        firestore.collection("/luanvan")
                .whereEqualTo("published",true).count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        published.setText(String.valueOf(task.getResult().getCount()));

                    }
                });
        firestore.collection("/luanvan")
                .whereEqualTo("published",false).count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        onprogress.setText(String.valueOf(task.getResult().getCount()));

                    }
                });

        firestore.collection("/tacgia").count()
                .get(AggregateSource.SERVER).addOnCompleteListener(new OnCompleteListener<AggregateQuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<AggregateQuerySnapshot> task) {
                        txtCCTongTG.setText(String.valueOf(task.getResult().getCount()));
                    }
                });

        firestore.collection("/linhvuc").document("/linhvuc")
                .addSnapshotListener(new EventListener<DocumentSnapshot>() {
                    @Override
                    public void onEvent(@Nullable DocumentSnapshot value, @Nullable FirebaseFirestoreException error) {
                        if(error!=null){

                        }
                        else{
                            if((List<String>)value.get("linhvuc")!=null){
                                txtCCTongCD.setText(String.valueOf(((List<String>)value.get("linhvuc")).size()));
                            }
                            refreshLayout.setRefreshing(false);
                        }
                    }
                });
    }


    private void AddView(View view) {
        historyBtn=view.findViewById(R.id.historyBtn);
        refreshLayout=view.findViewById(R.id.swiperLayout);
        mExit = (Button) view.findViewById(R.id.btnExit);

        all=view.findViewById(R.id.all);
        recent7=view.findViewById(R.id.recent7);
        recent30=view.findViewById(R.id.recent30);
        onprogress=view.findViewById(R.id.onprogress);
        published=view.findViewById(R.id.published);
        txtCCTongTG=view.findViewById(R.id.txtCCTongTG);
        txtCCTongCD=view.findViewById(R.id.txtCCTongCD);
        ViewInforBtn=view.findViewById(R.id.inforView);
        savetofileBtn=view.findViewById(R.id.savetofileBtn);
    }
    protected  void EventHandler(){
        refreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                DataHandler();
            }
        });
        historyBtn.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                getChildFragmentManager().beginTransaction()
                        .replace(R.id.congcuMain,new frag_lichsu())
                        .addToBackStack("xem lich su")
                        .commit();
            }
        });
        ViewInforBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                getChildFragmentManager().beginTransaction()
                        .replace(R.id.congcuMain,new frag_appinforView())
                        .addToBackStack("xem lich su")
                        .commit();
            }
        });
        savetofileBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                File path=Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
                File file=new File(path+"/luanvanText.txt");
                StringBuilder sb=new StringBuilder();
                firestore.collection("/luanvan").get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
                    @Override
                    public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                        List luanvan=new ArrayList(queryDocumentSnapshots.toObjects(LuanVan.class));

                        luanvan.forEach((lv)->{
                            sb.append(lv.toString());
                            try {
                                FileOutputStream fileOutputStream=new FileOutputStream(file);
                                fileOutputStream.write(sb.toString().getBytes());
                                fileOutputStream.close();
                                Toast.makeText(getContext(),"Lưu thông tin luận văn vào \nDownload/luanvanText.txt",Toast.LENGTH_LONG);


                            } catch (IOException e) {
                                Log.i("Lỗi đọc ghi file",e.toString());
                            }
                        });
                    }
                }).addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Toast.makeText(getContext(),"Lỗi xảy ra khi tải dữ liệu",Toast.LENGTH_SHORT);
                    }
                });
            }
        });
    }

    private LogoutListener logoutListener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof LogoutListener) {
            logoutListener = (LogoutListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement LogoutListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        logoutListener = null;
    }

    // When the button is clicked
    public void onLogoutButtonClicked() {
        if (logoutListener != null) {
            logoutListener.onLogout();
        }
    }


    private void Exit(View view) {
        mExit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
//                getActivity().finish();
                onLogoutButtonClicked();
            }
        });

    }





}

File: app/src/main/java/com/example/kmmkitchentmanagement/fragmenthome/frag_CongCu.kt
package com.example.kmmkitchentmanagement.fragmenthome

import android.content.Context
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.kmmkitchentmanagement.Model.LuanVan
import com.example.kmmkitchentmanagement.R
import com.example.kmmkitchentmanagement.fragmentSub.frag_appinforView
import com.example.kmmkitchentmanagement.fragmentSub.frag_lichsu
import com.example.kmmkitchentmanagement.interfaceFile.LogoutListener
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.*
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.util.*

class frag_CongCu : Fragment() {

    private lateinit var firestore: FirebaseFirestore
    private lateinit var refreshLayout: SwipeRefreshLayout
    private lateinit var all: TextView
    private lateinit var recent7: TextView
    private lateinit var recent30: TextView
    private lateinit var onprogress: TextView
    private lateinit var published: TextView
    private lateinit var txtCCTongTG: TextView
    private lateinit var txtCCTongCD: TextView
    private lateinit var mExit: Button
    private lateinit var historyBtn: Button
    private lateinit var ViewInforBtn: Button
    private lateinit var savetofileBtn: Button
    private var logoutListener: LogoutListener? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_frag__cong_cu, container, false)
        addView(view)
        exit(view)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        firebaseInit()
        dataHandler()
        eventHandler()
    }

    private fun firebaseInit() {
        firestore = FirebaseFirestore.getInstance()
    }

    private fun dataHandler() {
        val last7Day = Date().apply { date -= 7 }
        val last30Day = Date().apply { date -= 30 }

        firestore.collection("/luanvan").count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                all.text = task.result?.count.toString()
            }

        firestore.collection("/luanvan")
            .whereGreaterThan("addTime", last7Day)
            .count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                recent7.text = task.result?.count.toString()
            }

        firestore.collection("/luanvan")
            .whereGreaterThan("addTime", last30Day)
            .count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                recent30.text = task.result?.count.toString()
            }

        firestore.collection("/luanvan")
            .whereEqualTo("published", true)
            .count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                published.text = task.result?.count.toString()
            }

        firestore.collection("/luanvan")
            .whereEqualTo("published", false)
            .count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                onprogress.text = task.result?.count.toString()
            }

        firestore.collection("/tacgia").count()
            .get(AggregateSource.SERVER)
            .addOnCompleteListener { task ->
                txtCCTongTG.text = task.result?.count.toString()
            }

        firestore.collection("/linhvuc").document("/linhvuc")
            .addSnapshotListener { value, error ->
                if (error != null) {
                    return@addSnapshotListener
                }
                val linhvuc = value?.get("linhvuc") as? List<*>
                txtCCTongCD.text = linhvuc?.size.toString()
                refreshLayout.isRefreshing = false
            }
    }

    private fun addView(view: View) {
        historyBtn = view.findViewById(R.id.historyBtn)
        refreshLayout = view.findViewById(R.id.swiperLayout)
        mExit = view.findViewById(R.id.btnExit)
        all = view.findViewById(R.id.all)
        recent7 = view.findViewById(R.id.recent7)
        recent30 = view.findViewById(R.id.recent30)
        onprogress = view.findViewById(R.id.onprogress)
        published = view.findViewById(R.id.published)
        txtCCTongTG = view.findViewById(R.id.txtCCTongTG)
        txtCCTongCD = view.findViewById(R.id.txtCCTongCD)
        ViewInforBtn = view.findViewById(R.id.inforView)
        savetofileBtn = view.findViewById(R.id.savetofileBtn)
    }

    private fun eventHandler() {
        refreshLayout.setOnRefreshListener { dataHandler() }

        historyBtn.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .replace(R.id.congcuMain, frag_lichsu())
                .addToBackStack("xem lich su")
                .commit()
        }

        ViewInforBtn.setOnClickListener {
            parentFragmentManager.beginTransaction()
                .replace(R.id.congcuMain, frag_appinforView())
                .addToBackStack("xem lich su")
                .commit()
        }

        savetofileBtn.setOnClickListener {
            val path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
            val file = File(path, "luanvanText.txt")
            val sb = StringBuilder()

            firestore.collection("/luanvan").get()
                .addOnSuccessListener { queryDocumentSnapshots ->
                    val luanvan = ArrayList(queryDocumentSnapshots.toObjects(LuanVan::class.java))

                    luanvan.forEach { lv ->
                        sb.append(lv.toString())
                        try {
                            val fileOutputStream = FileOutputStream(file)
                            fileOutputStream.write(sb.toString().toByteArray())
                            fileOutputStream.close()
                            Toast.makeText(context, "Lưu thông tin luận văn vào \nDownload/luanvanText.txt", Toast.LENGTH_LONG).show()
                        } catch (e: IOException) {
                            Log.i("Lỗi đọc ghi file", e.toString())
                        }
                    }
                }
                .addOnFailureListener { e ->
                    Toast.makeText(context, "Lỗi xảy ra khi tải dữ liệu", Toast.LENGTH_SHORT).show()
                }
        }
    }

    private fun exit(view: View) {
        mExit.setOnClickListener {
            onLogoutButtonClicked()
        }
    }

    private fun onLogoutButtonClicked() {
        logoutListener?.onLogout()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is LogoutListener) {
            logoutListener = context
        } else {
            throw RuntimeException("$context must implement LogoutListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        logoutListener = null
    }
}
